generator/templates/wire/WireCmd.cpp

```diff
@@ -14,8 +14,250 @@
+{% macro serialize_member(member, in, out) %}
+    {%- if member.type.category == "object" -%}
+        {{out}} = provider.GetId({{in}});
+    {% elif member.type.category == "structure"%}
+        {{as_cType(member.type.name)}}Serialize({{in}}, &{{out}}, buffer, provider);
+    {%- else -%}
+        {{out}} = {{in}};
+    {%- endif -%}
+{% endmacro %}
+
+//* Outputs the deserialization code to put `in` in `out`
+{% macro deserialize_member(member, in, out) %}
+    {%- if member.type.category == "object" -%}
+        DESERIALIZE_TRY(resolver.GetFromId({{in}}, &{{out}}));
+    {% elif member.type.category == "structure"%}
+        DESERIALIZE_TRY({{as_cType(member.type.name)}}Deserialize(&{{out}}, &{{in}}, buffer, size, allocator, resolver));
+    {%- else -%}
+        {{out}} = {{in}};
+    {%- endif -%}
+{% endmacro %}
+
+//* The main [de]serialization macro
+
+//* Methods are very similar to structures that have one member corresponding to each arguments.
+//* This macro takes advantage of the similarity to output [de]serialization code for a record
+//* that is either a structure or a method, with some special cases for each.
+{% macro write_serialization_methods(name, members, as_method=None, as_struct=None, is_return_command=False) %}
+    {% set Return = "Return" if is_return_command else "" %}
+    {% set is_method = as_method != None %}
+    {% set is_struct = as_struct != None %}
+
+    //* Structure for the wire format of each of the records. Members that are values
+    //* are embedded directly in the structure. Other members are assumed to be in the
+    //* memory directly following the structure in the buffer.
+    struct {{name}}Transfer {
+        {% if is_method %}
+            //* Start the transfer structure with the command ID, so that casting to WireCmd gives the ID.
+            wire::{{Return}}WireCmd commandId;
+
+            //* Methods always have an implicit "self" argument.
+            ObjectId self;
+
+            //* Methods that return objects need to declare to the server which ID will be used for the
+            //* return value.
+            {% if as_method.return_type.category == "object" %}
+                ObjectId resultId;
+                ObjectSerial resultSerial;
+            {% endif %}
+        {% endif %}
+
+        //* Value types are directly in the command, objects being replaced with their IDs.
+        {% for member in members if member.annotation == "value" %}
+            {{member_transfer_type(member)}} {{as_varName(member.name)}};
+        {% endfor %}
+
+        //* const char* have their length embedded directly in the command.
+        {% for member in members if member.length == "strlen" %}
+            size_t {{as_varName(member.name)}}Strlen;
+        {% endfor %}
+    };
+
+    //* Returns the required transfer size for `record` in addition to the transfer structure.
+    size_t {{name}}GetExtraRequiredSize(const {{name}}& record) {
+        (void) record;
```

Comment:
Nit: if this is to silence an unused param warning, IWBN to have an UNUSED macro just to make it clearer.

Reply:
Done